### ISSUE TITLE
Sampled and regular softmax should use the same weight matrix shape

### ISSUE THREAD

**Issue by DavidNemeskey** [tensorflow/tensorflow#5350]
As it is now, the softmax samplers [sampled_softmax_loss](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#sampled_softmax_loss) and [nce_loss](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#nce_loss) require a `|C|x|H|` weight matrix, while the logits from [softmax](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#softmax) come from a `|H|x|C|`-sized one. This discrepancy is problematic on two levels:
1. it makes for an inconsistent API;
1. it results in a performance hit, because if one uses a sampled method for training and softmax for testing (as is usual when the number of classes is huge), one has to call `tf.transpose` somewhere, which does not work well with sparse input (see #4138).

In my case, I can choose between 21150 / 18300 or 22350 / 11900 train / test wps, depending on whether the `tf.transpose` happens on the sampled softmax loss during training or on regular softmax during testing. In either case, the performance is suboptimal.

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

As mentioned above, #4138. The fix there, however, is in the client code, not the API in question.

### Environment info
Operating System: `Linux 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u3 (2016-01-17) x86_64 GNU/Linux`

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
`/usr/local/cuda-8.0/lib64/libcudart.so.8.0.44`; I don't have cuDNN at the moment

If installed from binary pip package, provide:

1. A link to the pip package you installed: the official GPU install for Python 3.4
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`: `0.11.0rc1`

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
I cannot do that right now, but one can experiment with e.g. the PTB example.

### What other attempted solutions have you tried?
N/A

**Comment by aselle** [comment]
@lukaszkaiser, could you evaluate and decide whether this is something we want to do. It may be a difficult thing to change at this point given that the ubiquity of use of both of these. It may be worth providing a transposed version of softmax to allow people to optimize this case?

**Comment by DavidNemeskey** [comment]
@aselle I think out of the two, softmax has the "logical" dimensions, not the sampled functions. I preferred a new version / `transposed` parameter to the sampled functions to changing the softmax signature.

I also had a quick look at the code, and I am wondering if the problem might be that `embedding_lookup` is not able to perform a lookup on other than the first axis? I am not familiar with its implementation, but at least numpy can index along any of the axes, so it shouldn't be difficult to fix that.

**Comment by lukaszkaiser** [comment]
I did encounter this problem, and the transpose does somtimes incur cost. I don't know how large of a problem this is. I'm unassigning myself as I don't know the details of any softmax code, so I'm not sure how hard it'd be to provide an optional transpose argument.

**Comment by aselle** [comment]
Can you think of anybody who would be more appropriate?

**Comment by lukaszkaiser** [comment]
I think Stephan Gouws wrote the original NCE code, but I'm not sure who's working on it now, sorry.

**Comment by aselle** [comment]
@DavidNemeskey, I don't have anything concrete to share about sparse support right now. We are thinking about further sparse support more generally than just optimizers. Perhaps @gouwsmeister  could comment on his opinion.

**Comment by aselle** [comment]
@gouwsmeister, could you add anything further?

**Comment by gouwsmeister** [comment]
I did write the original code, but I have not kept up with the recent changes, so I may not be the best person to help with this. But I believe this was indeed originally because embedding_lookup indexes along the first dimension, as you mention. Also, to me CxH seems like the "logical" dimensions for the weight matrix, but I guess that's to some extent a matter of opinion :)

**Comment by vince62s** [comment]
Hey, I also encountered the issue  cf https://github.com/tensorflow/tensorflow/issues/5890
I really think this is a major issue since once you decide to use sampled_softmax for training (to be quick), of course you need the full softmax to be quick too at inference time.
Any update on this at the API level ?
Many thanks.

**Comment by b-3-n** [comment]
A simple workaround is to transpose the weight matrix once after training, assign it to a variable and use that so that the transpose is only done once.

**Comment by septra** [comment]
I agree with @DavidNemeskey here. I started working with tensorflow only recently and the discrepancy in the dimensions of the weight matrices passed to `sampled_softmax_loss` and `nce_loss` as compared to `softmax` was initially confusing. If it's not a priority to change the api, maybe a note in the documentation could be added about this?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by itsmeolivia** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
